### PR TITLE
Fix ldexp reverse-mode autodiff gradient at zero in tensorflow backend

### LINKED ISSUE #127244: tf.keras.ops.numpy.ldexp returns incorrect gradient at zero
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13.5

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.keras.ops.numpy.ldexp` returns an incorrect first derivative for a finite `float64` input.

In the reproducer below, the integer exponents are fixed and broadcast over the input. Therefore, `ldexp(x, e)` is just scaling by fixed powers of two, and the weighted result is affine in `t`. Zero is not a singularity of scaling by a fixed power of two.

The forward value is computed correctly as `101.5`, but reverse-mode autodiff returns `209.0` for the derivative with respect to `t`. The expected derivative is `253.125`.

Expected behavior?

The first derivative of the weighted `tf.keras.ops.numpy.ldexp` expression at `t = 0.0` should be close to `253.125`, not `209.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    x = tf.reshape(tf.stack([t, 2 * t + 1]), (2, 1))
    e = tf.constant([-3, 0, 4], dtype=tf.int32)
    y = tf.keras.ops.numpy.ldexp(x, e)
    return sum((1 + 3 * i + j) * y[i, j] for i in range(2) for j in range(3))

x = tf.constant(0.0, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 253.125

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.keras.ops.numpy.ldexp returns incorrect gradient at zero")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 101.5
actual: 209.0
expected: 253.125
BUG REPRODUCED: tf.keras.ops.numpy.ldexp returns incorrect gradient at zero
```

### PR DESCRIPTION
**Pull request by jdymitarai**
## Description

In `keras/src/backend/tensorflow/numpy.py`, `ldexp` previously used:
```python
result = x1 * tf.pow(tf.constant(2.0, dtype=x1.dtype), x2)
return tf.cast(tf.where(tf.math.is_inf(x1) | (x1 == 0), x1, result), dtype)
```

When `x1 == 0`, `tf.where` selected the `x1` branch. In reverse-mode autodiff, differentiating `x1` evaluates to $1.0$ rather than the expected scaled power of two $2^{x2}$. This caused gradients at $x1 = 0$ to be incorrect (as reported in tensorflow/tensorflow#127244).

Because $0.0 \times 2^{x2}$ naturally evaluates to $0.0$, the `| (x1 == 0)` check is unnecessary. Only `tf.math.is_inf(x1)` needs to select `x1` to prevent $\pm\infty \times 0 \to \text{NaN}$.

### Fix
1. Removed `| (x1 == 0)` in `backend/tensorflow/numpy.py`.
2. Added zero-input tests in `keras/src/ops/numpy_test.py`.

Fixes tensorflow/tensorflow#127244

## Contributor Agreement

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #xxx` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Comment by jdymitarai** [comment]
Updated the implementation and tests based on code review:
1. Preserve signed zero (\-0.0\) by returning \x1\ instead of \zeros_like(x1)\ when \x1 == 0\ and \pow_val\ is infinite.
2. Added comprehensive unit tests in float64 (\x2 = 1024\) and float32 (\x2 = 128\) to verify that both \+0.0\ and \-0.0\ directly paired with overflowing exponents never produce \NaN\ and strictly preserve their IEEE 754 signbit.

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!

**Comment by hertschuh** [comment]
Are you able to also fix the OpenVino implementation? It looks like it has the same issue.
